### PR TITLE
Improve performance of Icinga 2 API commands processed in bulk

### PR DESCRIPTION
**Pull request by Al2Klimov**
Let me explain how I came here in the first place.

While discussing removal of some deprecated features with @Thomas-Gelf he opted against removing the command pipe. He considered it faster than the Icinga 2 API for massive stuff submission. I thought: wait, we have one system which eventually does massive stuff submission via API – Icinga Web. I even actually did that while testing Icinga ack sync. And it was indeed slow. I asked myself: what could it already do by itself to do this faster? What does it do right now? While reverse-engineering this repo I came to the conclusion that it renders, sends and awaits commands one by one. I remembered that @yhabteab did somewhere something with PHP event loops and let him show me them. They could help me sending multiple commands concurrently. It turned out more complicated than I thought, but then I remembered something else. @julianbrost digged out than our API supports not only DSL filters and filter_vars which iterate over all objects. It also supports targeting a specific object by name. This repo already does this. But it supports targeting multiple specific objects by array of names as well. That's what I did here.

* Icinga/icinga2#6307
* Icinga/icinga2#9652
* Icinga/icinga2#9188

Now the command forms render at best exactly one Icinga 2 API action request targeting all selected objects at once, not one request per object. Commenting e.g. 500 hosts now takes 4s, not 40. For this now commands and forms always handle multiple objects by design. PHP generators everywhere don't permit the RAM to blow up. Delete*Command are now parts of the ObjectsCommand family.

## TODO

* [x] Approval of the concept in general
* [x] Agreement on the implementation details
* [x] What's up with GHA? (@Al2Klimov)
* [x] Tests of all code usages of all commands DB-Web supports

### CODE REVIEW

File: application/forms/Command/Object/ScheduleCheckForm.php [hunk 1]
```diff
@@ -109,26 +111,25 @@ protected function assembleSubmitButton()
...
-        }
+        list($firstGranted, $allGranted) = $this->copyFirstItem(new CallbackFilterIterator(
+            $objects instanceof Iterator ? $objects : new IteratorIterator($objects),
+            function (Model $object): bool {
+                return $this->isGrantedOn('icingadb/command/schedule-check', $object) || (
+                    $object->active_checks_enabled
+                    && $this->isGrantedOn('icingadb/command/schedule-check/active-only', $object)
+                );
+            }
+        ));
```
Comment (Al2Klimov) [inline application/forms/Command/Object/ScheduleCheckForm.php:124]: ... **and** a whiteboard for this. 😅

8F58897B-41A5-4D25-8BE4-E8A6FFB17186

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Icingadb/Command/Object/ObjectsCommand.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     * Involved objects
+     *
+     * @var Model[]
+     */
+    protected $objects;
+
+    /**
+     * Set the involved objects
+     *
+     * @param   Model[] $objects
```
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Command/Object/ObjectsCommand.php:26]: I'm afraid the declaration is pretty limited on this:

https://stackoverflow.com/a/34273821

[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Command/Object/ObjectsCommand.php:26]: I know. But you shouldn't deviate from the declared argument type in the docstring. That's not just me who's not liking it, https://github.com/Icinga/icingadb-web/pull/423 won't as well.

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Command/Object/ObjectsCommand.php:26]: I.e. I shall only advertise an array (of T) if I declare an (even untyped) `array`?

[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Command/Object/ObjectsCommand.php:26]: Yes, either remove the declaration or document the type as `Traversable`

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Command/Object/ObjectsCommand.php:26]: I think there's a misunderstanding. I meant neither, but to document `T[]`, but to PHP-declare `array` as PHP can't `T[]` (only PHP**Doc**):

> I'm afraid the declaration is pretty limited on this:
> 
> https://stackoverflow.com/a/34273821

OK?

[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Command/Object/ObjectsCommand.php:26]: You didn't want an array in the first place. So why not:

```
/**
* @param Model[] $objects
*/
public function setObjects($objects)
```

or

```
/**
* @param Traversable $objects
*/
public function setObjects(Traversable $objects)
```

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply library/Icingadb/Command/Object/ObjectsCommand.php:26]: > You didn't want an array in the first place.

I changed my mind, especially after https://github.com/Icinga/icingadb-web/pull/721#discussion_r1259558630 . Arrays are harder to screw up, so I chose to PHP-enforce `array`. OK?

[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Icingadb/Command/Object/ObjectsCommand.php:26]: If you enforce `array`, you prevent the usage of database queries and require everyone passing the result of a `$query->fetchAll()` blowing up the memory usage. That's not OK.

File: application/forms/Command/CommandForm.php [hunk 3]
```diff
@@ -133,4 +128,50 @@ protected function sendCommand(IcingaCommand $command)
...
+     *
+     * @return Generator
+     */
+    protected function forwardAllItems(Iterator $it): Generator
+    {
+        do {
+            yield $it->current();
+            $it->next();
+        } while ($it->valid());
+    }
```
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply application/forms/Command/CommandForm.php:176]: So far I think I've fixed everything ex. this. Are you sure we wouldn't violate any interface rules like 1) rewind 2) valid ... here?

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply application/forms/Command/CommandForm.php:176]:
```
php > function g() { yield 1; yield 2; }
php > $x=g();
php > var_dump($x->valid());
bool(true)
php > var_dump($x->rewind());
NULL
php > var_dump($x->valid());
bool(true)
php > var_dump($x->current());
int(1)
php >
```

OK... :D

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply application/forms/Command/CommandForm.php:176]: Oh, almost forgot:


```
php > function g() { if (0) yield 1; }
php > $x=g();
php > var_dump($x->valid());
bool(false)
php >
```

For the record.

[the remark this answers is not in the document]
Comment (nilmerg) [reply application/forms/Command/CommandForm.php:176]: Then don't use it:

```php
$objects = function ($it) {
    foreach ($it as $o) {
        if (...) {
            yield $o;
        }
    }
}();

```

File: application/forms/Command/Object/CheckNowForm.php [hunk 4]
```diff
@@ -4,17 +4,19 @@
...
+use Generator;
 use Icinga\Module\Icingadb\Command\Object\ScheduleCheckCommand;
-use Icinga\Module\Icingadb\Common\Auth;
 use Icinga\Module\Icingadb\Forms\Command\CommandForm;
 use Icinga\Web\Notification;
 use ipl\Orm\Model;
 use ipl\Web\Widget\Icon;
+use Iterator;
+use IteratorIterator;
+use Traversable;
```
Comment (nilmerg) [inline application/forms/Command/Object/CheckNowForm.php:16]: Cleanup unused usages.

File: application/forms/Command/Object/DeleteCommentForm.php [hunk 5]
```diff
@@ -4,17 +4,20 @@
...
 use Icinga\Module\Icingadb\Command\Object\DeleteCommentCommand;
-use Icinga\Module\Icingadb\Common\Auth;
 use Icinga\Module\Icingadb\Forms\Command\CommandForm;
 use Icinga\Web\Notification;
 use ipl\Orm\Model;
 use ipl\Web\Common\RedirectOption;
 use ipl\Web\Widget\Icon;
+use Iterator;
+use IteratorIterator;
+use Traversable;
```
Comment (nilmerg) [inline application/forms/Command/Object/DeleteCommentForm.php:17]: Cleanup unused usages.

File: application/forms/Command/Object/DeleteDowntimeForm.php [hunk 6]
```diff
@@ -4,17 +4,20 @@
...
 use Icinga\Module\Icingadb\Command\Object\DeleteDowntimeCommand;
-use Icinga\Module\Icingadb\Common\Auth;
 use Icinga\Module\Icingadb\Forms\Command\CommandForm;
 use Icinga\Web\Notification;
 use ipl\Orm\Model;
 use ipl\Web\Common\RedirectOption;
 use ipl\Web\Widget\Icon;
+use Iterator;
+use IteratorIterator;
+use Traversable;
```
Comment (nilmerg) [inline application/forms/Command/Object/DeleteDowntimeForm.php:17]: Cleanup unused usages.

File: application/forms/Command/Object/ProcessCheckResultForm.php [hunk 7]
```diff
@@ -4,8 +4,9 @@
 
 namespace Icinga\Module\Icingadb\Forms\Command\Object;
 
+use CallbackFilterIterator;
```
Comment (nilmerg) [inline application/forms/Command/Object/ProcessCheckResultForm.php:7]: Cleanup unused usages.

File: application/forms/Command/Object/ScheduleCheckForm.php [hunk 8]
```diff
@@ -17,11 +18,12 @@
 use ipl\Orm\Model;
 use ipl\Web\FormDecorator\IcingaFormDecorator;
 use ipl\Web\Widget\Icon;
+use Iterator;
+use IteratorIterator;
+use Traversable;
```
Comment (nilmerg) [inline application/forms/Command/Object/ScheduleCheckForm.php:23]: Cleanup unused usages.

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Comment by Al2Klimov** [comment]
# Tests

## To do list

```
$ git diff --stat master..bulk-cmd application/forms/Command
 application/forms/Command/CommandForm.php                         | 51 +++++++++++++++++++++++++++++++--------------------
 application/forms/Command/Instance/ToggleInstanceFeaturesForm.php |  7 ++-----
 application/forms/Command/Object/AcknowledgeProblemForm.php       | 42 ++++++++++++++++++------------------------
 application/forms/Command/Object/AddCommentForm.php               | 34 ++++++++++++++--------------------
 application/forms/Command/Object/CheckNowForm.php                 | 45 +++++++++++++++++++++++----------------------
 application/forms/Command/Object/DeleteCommentForm.php            | 30 ++++++++++++++++--------------
 application/forms/Command/Object/DeleteDowntimeForm.php           | 36 +++++++++++++++++++-----------------
 application/forms/Command/Object/ProcessCheckResultForm.php       | 37 ++++++++++++++++++-------------------
 application/forms/Command/Object/RemoveAcknowledgementForm.php    | 24 +++++++++---------------
 application/forms/Command/Object/ScheduleCheckForm.php            | 45 +++++++++++++++++++++++----------------------
 application/forms/Command/Object/ScheduleHostDowntimeForm.php     | 53 +++++++++++++++++++++++++----------------------------
 application/forms/Command/Object/ScheduleServiceDowntimeForm.php  | 44 +++++++++++++++++++-------------------------
 application/forms/Command/Object/SendCustomNotificationForm.php   | 28 +++++++++++-----------------
 application/forms/Command/Object/ToggleObjectFeaturesForm.php     | 32 +++++++++++++++-----------------
 14 files changed, 243 insertions(+), 265 deletions(-)
$
```

## Instance/ToggleInstanceFeaturesForm.php

Bildschirmaufnahme 2023-08-02 um 12 46 39

## Object/AcknowledgeProblemForm.php

Bildschirmaufnahme 2023-08-02 um 12 57 32

## Object/AddCommentForm.php

Bildschirmaufnahme 2023-08-02 um 12 59 25

## Object/CheckNowForm.php

Bildschirmaufnahme 2023-08-02 um 13 01 16

## Object/DeleteCommentForm.php

Bildschirmaufnahme 2023-08-02 um 13 03 06

## Object/ScheduleServiceDowntimeForm.php

Bildschirmaufnahme 2023-08-02 um 13 04 02

## Object/DeleteDowntimeForm.php

Bildschirmaufnahme 2023-08-02 um 13 05 00

## Object/RemoveAcknowledgementForm.php

Bildschirmaufnahme 2023-08-02 um 13 06 22

## Object/ProcessCheckResultForm.php

Bildschirmaufnahme 2023-08-02 um 13 07 04

## Object/ToggleObjectFeaturesForm.php

Bildschirmaufnahme 2023-08-02 um 13 07 23

## Object/ScheduleCheckForm.php

Bildschirmaufnahme 2023-08-02 um 13 09 18

## Object/SendCustomNotificationForm.php

Bildschirmaufnahme 2023-08-02 um 13 10 54

## Conclusion

It still works.

**Review by nilmerg — APPROVED** [review]

**Pull request merged**